fdbserver/include/fdbserver/DataDistribution.actor.h

```diff
@@ -340,6 +342,154 @@ struct ShardTrackedData {
 	Reference<AsyncVar<Optional<ShardMetrics>>> stats;
 };
 
+// Holds the permitted size and IO Bounds for a shard
+struct ShardSizeBounds {
+	StorageMetrics max;
+	StorageMetrics min;
+	StorageMetrics permittedError;
+
+	bool operator==(ShardSizeBounds const& rhs) const {
+		return max == rhs.max && min == rhs.min && permittedError == rhs.permittedError;
+	}
+};
+
+// Gets the permitted size and IO bounds for a shard
+ShardSizeBounds getShardSizeBounds(KeyRangeRef shard, int64_t maxShardSize);
+
+// Determines the maximum shard size based on the size of the database
+int64_t getMaxShardSize(double dbSizeEstimate);
+
+#ifndef __INTEL_COMPILER
+#pragma endregion
+#endif
+
+/////////////////////////////// DD Components //////////////////////////////////////
+#ifndef __INTEL_COMPILER
+#pragma region DD Components
+#endif
+class DDTeamCollection;
+
+// send request/signal to TeamCollection through interface
+// call synchronous method from components outside DDTeamCollection
+struct ITeamCollection {
+	struct Interface {
+		PromiseStream<GetTeamRequest> getTeam;
+	};
+	virtual ~ITeamCollection() {}
+};
+
+// send request/signal to DDTracker through interface
+// call synchronous method from components outside DDTracker
+struct IDDTracker {
+	struct Interface {
+		Promise<Void> readyToStart;
+		PromiseStream<GetMetricsRequest> getShardMetrics;
+		PromiseStream<GetTopKMetricsRequest> getTopKMetrics;
+		PromiseStream<GetMetricsListRequest> getShardMetricsList;
+		PromiseStream<KeyRange> restartShardTracker;
+		PromiseStream<Promise<int64_t>> getAverageShardBytes; // FIXME(xwang): change it to a synchronous call
+	};
+	virtual double getAverageShardBytes() = 0;
+	virtual ~IDDTracker() = 0;
+};
+
+// send request/signal to DDQueue through interface
+// call synchronous method from components outside DDQueue
+struct IDDQueue {
```

Comment:
Might be good to have these be in separate header file specific to queue, tracker etc

Reply:
I have a second through actually. In my opinion, I think the Interface and DDContext should be in `DataDistribution.actor.h` because they will be visible to every component in DD. By contrast, the implementation of those interfaces (like if DDQueue implements IDDQueue) should be in a separate file.

Reply:
DDContext, maybe, but code that doesn't need to interact with the tracker shouldn't have to include everything. Having large all-inclusive header files also results in longer compilation times